!!! FEATURE: Extract workspace metadata and user-assignment to Neos

bwaidelich opened the pull request:
Introduces `WorkspacePublishingService` as replacement for the current Neos `Workspace` "active record" model.

Introduces `WorkspaceService` as central authority to manage Neos workspaces.

## Migration

To migrate existing workspace metadata & roles, run

```
./flow doctrine:migrate
./flow migrateevents:migrateWorkspaceMetadataToWorkspaceService
```

(The doctrine migrations definitely have to be executed before the first login!)

Related: #4726
Related: https://github.com/neos/neos-development-collection/pull/4186 (introduced workspace renaming etc)

bwaidelich committed:
FEATURE: Extract workspace metadata and user-assignment to Neos

Introduces `WorkspacePublishingService` as replacement for the current Neos `Workspace` "active record" model.

Introduces `WorkspaceService` as central authority to manage Neos workspaces.

Related: #4726

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,349 @@
+        ContentRepositoryId $contentRepositoryId,
+        WorkspaceTitle $title,
+        WorkspaceDescription $description,
+    ): WorkspaceName
+    {
```
```suggestion
    ): WorkspaceName {
```

bwaidelich replied:
Thanks, but this is just a draft. Please don't add comments re cosmetics just yet

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
```diff
@@ -81,6 +81,14 @@ public function get(ContentRepositoryId $contentRepositoryId): ContentRepository
+    /**
+     * @return array<ContentRepositoryId>
+     */
+    public function getContentRepositoryIds(): array
```
thanks i also need this for listing all content repositories via cli or for example for the new 9.0 neos setup :)

mhsdesign commented on Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php:
```diff
@@ -61,19 +68,16 @@ public static function forLive(): self
-        try {
-            // Check if name already match name pattern to prevent unnecessary transliteration
+        if (self::hasValidFormat($name)) {
             return self::fromString($name);
-        } catch (\InvalidArgumentException $e) {
-            // Okay, let's transliterate
         }
```
❤️ this part always bummed me :D

mhsdesign commented on Neos.Neos/Classes/Command/WorkspaceCommandController.php:
```diff
@@ -178,13 +179,23 @@ public function createCommand(
+
+        $this->workspaceService->createWorkspace(
+            $contentRepositoryId,
+            \Neos\Neos\Domain\Model\WorkspaceTitle::fromString($title),
```
should probably be `use`d instead of fq

bwaidelich replied:
This is a temporary work around, as the `\Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceTitle` will be replaces in the next step.
But you're right, I should probably rather import the _new_ namespace if the other one is to be removed.

This is just a draft though

mhsdesign commented on Neos.Neos/Classes/Command/WorkspaceCommandController.php:
```diff
@@ -359,16 +367,18 @@ public function listCommand(string $contentRepository = 'default'): void
 
         $tableRows = [];
-        $headerRow = ['Name', 'Base Workspace', 'Title', 'Owner', 'Description', 'Status', 'Content Stream'];
+        $headerRow = ['Name', 'Classification', 'Base Workspace', 'Title', 'Description', 'Status', 'Content Stream'];
```
so we dont want to show the user id? Well i guess that makes sense as its too technical anyway and the human readable name with classification personal will already suffice?

bwaidelich replied:
I'll certainly re-add it, but with the new Workspace <-> Role assignment, maybe rather in some new `showCommand()` that lists all contributors and their role

mhsdesign commented on Neos.Neos/Classes/Domain/Model/User.php:
```diff
@@ -36,6 +36,11 @@ class User extends Person implements UserInterface
      */
     protected $preferences;
 
+    /**
+     * @var string
+     */
+    protected $Persistence_Object_Identifier;
```
```suggestion
    /**
     * This property will be introduced via Flows persistence magic aspect
     * We declare it here already as workaround to not have to use {@see PersistenceManager::getIdentifierByObject()} to access this identifier.
     * @var string
     */
    protected $Persistence_Object_Identifier;
```

nice "hack" idea ... :D I approve ... probably better than injecting the persistence manager to do the same? But the above property should get a nice doc comment

mhsdesign commented on Neos.Neos/Classes/Domain/Model/UserId.php:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @api
+ */
+final readonly class UserId implements \JsonSerializable
```
will this replace the cr's user id thing?

bwaidelich replied:
It probably won't _replace_ it because the CR also has a concept of users (currently only used to store the `initiatingUserId` in the event metadata).
But IMO it makes sense to have a dedicated "Neos UserId" as well.. Maybe we can rename one to make the distinction clearer..

mhsdesign reviewed (commented):
Some early thoughts thanks for the progress so far ❤️ 

Notes from my side:
- [x] remove `\Neos\Neos\Domain\Service\UserService::deletePersonalWorkspace` and `\Neos\Neos\Domain\Service\UserService::removeOwnerFromUsersWorkspaces`
- [x] deprecate `\Neos\Neos\Service\UserService::getPersonalWorkspaceName`
- [x] remove `WorkspaceNameBuilder` again

bwaidelich commented on Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php:
```diff
@@ -84,7 +88,7 @@ public static function transliterateFromString(string $name): self
         }
 
-        return new self($name);
+        return self::fromString($name);
```
TODO: check max length and truncate if too long!

bwaidelich replied:
That's already enforced with line 91ff

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
cc @Sebobo there are quite some changes in this workspace controller ... just to inform you as youre working on this as well :D

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,349 @@
+            ),
+            WorkspaceClassification::ROOT => WorkspacePermissions::create(
+                read: true,
+                publish: $this->privilegeManager->isPrivilegeTargetGrantedForRoles($userRoles, 'Neos.Workspace.Ui:Backend.PublishAllToLiveWorkspace'),
```
i think the privilge should belong to neos/neos or neos/neos-core still and not be moved via https://github.com/neos/neos-development-collection/pull/5118

mhsdesign replied:
or as alternative of moving the `Neos.Workspace.Ui:Backend.PublishAllToLiveWorkspace` target back to neos, or creating a new empty target for this which will be interpreted here, we could check for the `LivePublisher` role?

`$this->securityContext->hasRole('Neos.Neos:LivePublisher')`

edit: we cant move the target back to `Neos.Neos:Backend.PublishAllToLiveWorkspace` because it protects `Ui\Controller\WorkspaceController->publishWorkspaceAction` which is in the other package.

bwaidelich replied:
yes, I was wondering the same.. There is some mismatch between workspace <-> user/role assignment and the global Flow roles & privileges..
I'll think about it tomorrow!

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
```diff
@@ -81,6 +81,14 @@ public function get(ContentRepositoryId $contentRepositoryId): ContentRepository
         return $this->getFactory($contentRepositoryId)->getOrBuild();
     }
 
+    /**
+     * @return array<ContentRepositoryId>
+     */
+    public function getContentRepositoryIds(): array
+    {
+        return array_map(ContentRepositoryId::fromString(...), array_keys($this->settings['contentRepositories']));
+    }
```
```suggestion
    /**
     * @return iterable<ContentRepositoryId>
     */
    public function getContentRepositoryIds(): iterable
    {
        return array_map(ContentRepositoryId::fromString(...), array_keys($this->settings['contentRepositories'] ?? []));
    }
```

this way we can easily introduce a collection dto at some point and it wont crash if nothing is configured.

bwaidelich replied:
Resolved with #5194

mhsdesign commented on Neos.Neos/Classes/Domain/Model/UserInterface.php:
Fyi since you wondered and i stumbled upon it ... we have this interface for the following reason:

https://github.com/neos/neos-development-collection/pull/165#issuecomment-157645872


and actually this thing is now useless with the new cr.

mhsdesign replied:
deprecated it now

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
wip

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Mini tweaks

bwaidelich committed:
wip

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/Workspace/Workspace.php:
```diff
@@ -24,31 +24,72 @@
      */
     public const PERSONAL_WORKSPACE_PREFIX = 'user-';
 
+    /**
```
Note: The following properties have been turned into actual fields to improve IDE support:
Constructor promoted arguments of an `@internal` constructor are all "internal", but in this case they are _not_. Besides the `@deprecated` tags are ignored otherwise

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,316 @@
+        $workspaceNameCandidate = WorkspaceName::transliterateFromString($candidate);
+        $workspaceName = $workspaceNameCandidate;
+        while ($contentRepository->getWorkspaceFinder()->findOneByName($workspaceName) instanceof Workspace) {
+            $workspaceName = WorkspaceName::fromString(
```
with the lowercase enforcement `generateRandomString` is not safe and will crash

bwaidelich replied:
This should be resolved with 5f1e296ca123788d14beb2162c50431ddee2551d

mhsdesign committed:
WIP: Fix WorkspaceService as discussed with basti & christian

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Fix WorkspaceService::createRootWorkspace()

bwaidelich committed:
Use `WorkspaceService` in workspace::createroot CLI command

bwaidelich committed:
Stabilize unique workspace name creation

and add initial behat tests

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Provide CLI to synchronize workspace metadata and roles

Usage:

```
./flow workspace:syncall --content-repository foo
```

bwaidelich committed:
Stabilize and document WorkspaceService

bwaidelich committed:
Fix workspace role handling

bwaidelich committed:
Improve "pending changes" projection and read model

bwaidelich committed:
Cosmetic tweaks to satisfy linter

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Move owner assignment from role to metadata, fix migration

bwaidelich committed:
Remove UI specific publishable node info array from `Changes` VO

bwaidelich committed:
Make `WorkspaceNameBuilder` obsolete and remove it

bwaidelich committed:
Remove workspace related left-overs from `UserService`

mhsdesign commented on Neos.Neos/Classes/Utility/User.php:
```diff
@@ -16,13 +18,20 @@ class User
     public static function getPersonalWorkspaceNameForUsername($username): string
     {
-        return WorkspaceNameBuilder::fromAccountIdentifier($username)->value;
+        $name = preg_replace('/[^A-Za-z0-9\-]/', '-', 'user-' . $username);
```
isnt this logic now completely outdated? i think we HAVE to tell users to use `createPersonalWorkspaceForUserIfMissing` (and thus `findPrimaryWorkspaceNameForUser`) this static utility is now obsolete.

bwaidelich replied:
It's true.. Currently this is still used by the demo site (https://github.com/neos/Neos.Demo/blob/09864afab88b715106232b8769d6359e04cf1b3f/Classes/Form/Runtime/Action/CreateUserAction.php#L65) but we can probably replace that

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/Workspace/Workspace.php:
```diff
@@ -24,31 +24,72 @@
+     */
+    public WorkspaceStatus $status;
+
+    /**
+     * @deprecated with 9.0.0-beta12 owners/collaborators should be assigned to workspaces outside the Content Repository core
+     */
+    public string|null $workspaceOwner;
```
did that even work before ... and how does the deprecation strategy look like? Can we just not remove it with the migration in place?

mhsdesign replied:
Also `CreateWorkspace:$workspaceOwner` is not deprecated?

bwaidelich replied:
> Can we just not remove it with the migration in place?

What do you mean?

> Also CreateWorkspace:$workspaceOwner is not deprecated?

Right, that has to be deprecated. On it

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/Workspace/Workspace.php:
```diff
@@ -24,31 +24,72 @@
     /**
      * This prefix determines if a given workspace (name) is a user workspace.
      */
     public const PERSONAL_WORKSPACE_PREFIX = 'user-';
```
this can also be removed or deprecated?

bwaidelich replied:
good call, I'll move it to the only usage (`Workspace::isPersonalWorkspace()`)

mhsdesign reviewed (changes requested):
I think we still need to draw out a good strategy how the `Workspace` model should look like in the final release and how we can migrate the deprecated things and remove them.

bwaidelich committed:
Make `Utility\User` class obsolete and remove it

bwaidelich committed:
Rework migration to work on the database table directly

mhsdesign committed:
TASK: Make migration more stable by not relying on the projection but calculating the state via events

bwaidelich committed:
Optimize `migrateWorkspaceMetadataToWorkspaceService` migration

by only loading relevant events

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

mhsdesign committed:
TASK: Adjust deprecations

mhsdesign committed:
TASK: Deprecate `UserInterface`

The interface is now obsolete with the ESCR and was previously only used as pragmatic solution to decouple the CR from Neos.Neos

as written here:

> Some additional background information: This PR also introduces a new "owner" field to the `Workspace` model in TYPO3CR. The owner of a workspace is a Neos user. But in order to avoid a dependency from TYPO3CR to Neos, I simply store a UUID and introduced a new `UserInterface` in TYPO3CR. [...]

see https://github.com/neos/neos-development-collection/pull/165#issuecomment-157645872

mhsdesign committed:
TASK: Add comment for `$Persistence_Object_Identifier` "hack"

The property will be introduced via Flows persistence magic aspect.
We declare it here already as workaround to not have to use `PersistenceManager::getIdentifierByObject()`}` to access this identifier.
And to have IDE support.

mhsdesign committed:
TASK: Improve exceptions

mhsdesign committed:
TASK: Make migration independent of workspaceService

Otherwise, we would be implicitly still relying on the workspace finder and up-to-date projections

mhsdesign committed:
TASK: Add missing `@internal` annotations

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspaceSubjectType.php:
```diff
@@ -0,0 +1,14 @@
+/**
+ * @api
+ */
+enum WorkspaceSubjectType : string
```
id favour a doc comment here as well how this works :D

bwaidelich replied:
yes, good call

mhsdesign committed:
TASK: Remove obsolete todos

(There is no acl to bypass yet *g)

mhsdesign reviewed (approved):
Thank you so much for pushing this till the end!
We discussed the future adjustments - what should happen with the `@deprecated` things, and how will the workspace be part of the content graph projection - and this will all be part - as originally anticipated but nearly forgotten - of https://github.com/neos/neos-development-collection/pull/5096

We made some final adjustments together so im happy :D And we also tested the migration.
I just skimmed again through all the code (some part more extensively some parts rather quickly) but it looks like a good step in total.

After some additional adjustments the Neos Ui e2e test are recovering. But they are sadly out of order partly on 9.0 so we wont get total confirmation from that.

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+    /**
+     * Load metadata for the specified workspace
+     */
+    public function getWorkspaceMetadata(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName): WorkspaceMetadata
```
since we are also at `getWorkspacePermissionsForUser` forgiving i would also not throw here at `getWorkspaceMetadata` ... dindt we introduce `WorkspaceClassification::UNKNOWN` for that reason?

bwaidelich replied:
yes, I agree that this method could always return metadata. `WorkspaceClassification::UNKNOWN` is a bit weird if we use that enum to _create_ or _change_ the classification – but maybe we never need it then (or disallow that enum unit in those cases)

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspacePermissions.php:
```diff
@@ -0,0 +1,39 @@
+final readonly class WorkspacePermissions
+{
+    private function __construct(
+        public bool $read,
+        public bool $write,
+        public bool $manage,
```
what is the meaning of `false, false, true` (if the user is admin he can manage ... but what is manage again? ^^) i think we need here a doc comment as well *g

bwaidelich replied:
manage = change the workspace roles/metadata itself. But, yes, I'll add doc comments

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspaceClassification.php:
```diff
@@ -0,0 +1,17 @@
+    case SHARED = 'SHARED';
+    case ROOT = 'ROOT';
+
+    case UNKNOWN = 'UNKNOWN';
```
currently unused

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+        if (!is_array($metadataRow)) {
+            return null;
+        }
+        assert(is_string($metadataRow['title']));
+        assert(is_string($metadataRow['description']));
+        assert(is_string($metadataRow['classification']));
+        assert(is_null($metadataRow['owner_user_id']) || is_string($metadataRow['owner_user_id']));
```
i dont get the sense behind this ... php would just throw an exception either way? And that is likely more descriptive than the error here. I would just rather not have this logic here duplicated for the case it being faulty.

bwaidelich replied:
without those, PHPStan will complain

bwaidelich replied:
no it won't :)
I'll remove it

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+     */
+    public function getWorkspacePermissionsForUser(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, User $user): WorkspacePermissions
+    {
+        $userIsAdministrator = $this->userService->currentUserIsAdministrator();
```
we should probably at some point have a different logic for that than to use the flow `Neos.Neos:Administrator` right? But maybe not either ... so doesnt matter... but i was thinking about a privilege target that the admin has and that you can add to your own `CRAdministrator` role.

bwaidelich replied:
I don't understand

mhsdesign replied:
we discussed that this is perfekt and a `My.Vendor:CRAdministrator` manager permission can be added by adding a `MANAGER` entry for every workspace to the `neos_neos_workspace_role` (api yet to be created)

mhsdesign reviewed (commented):
I found some last inconsistencies... after having trouble to run the neos ui e2e tests as we now have to call the migrate script:

```
./flow cr:setup --content-repository twodimensions
./flow cr:import --content-repository twodimensions --path ./DistributionPackages/Neos.Test.TwoDimensions/Resources/Private/Content
# Hack to create Neos workspace metadata for the live workspace (Failed to load metadata for workspace "live").
./flow migrateevents:migrateWorkspaceMetadataToWorkspaceService --content-repository onedimension
```

... which also only works because they are old events. So if you would have just an new eventstream with the live workspace obviously there is no way to repair neos besides creating the table entries by hand.

So i think we either should have

A: a command which inserts dummy metadata for missing workspaces

B: return dummy metadata at runtime to prevent an exception


we are already forgiving regarding  `getWorkspacePermissionsForUser` but as currently implemented it would restrict write access to the say live workspace meaning one has a working neos at least but noone can publish to live.

The obvious long term solution is to have a good in & exporter taking care of that but we dont have that luxury right now and i think there should be a way to setup a Neos just by having the live workspace event stream and the source code. Everything else must not be that critical. But i might also be alone with that assumption.

mhsdesign commented on Neos.Neos/Classes/Service/UserService.php:
```diff
@@ -61,33 +59,6 @@ public function getBackendUser()
-     *
-     * @api
-     */
-    public function getPersonalWorkspaceName(): ?string
```
despite that being earlier available in `context.userInformation.personalWorkspaceName` in security i think so too that this can go? Especially since security will work differently in Neos 9 right?

bwaidelich replied:
right, I agree

mhsdesign commented on Neos.Neos/Classes/Domain/Service/UserService.php:
```diff
@@ -247,39 +247,21 @@ public function getCurrentUser(): ?User
-
-    public function getCurrentUserIdentifier(): ?UserId
-    {
-        $currentUser = $this->getCurrentUser();
-
-        return $currentUser
-            ? UserId::fromString($this->persistenceManager->getIdentifierByObject($currentUser))
-            : null;
+        }
+        return null;
     }
 
-    public function findByUserIdentifier(UserId $userId): ?User
```
just checked, its not breaking to rename this one and remove `getCurrentUserIdentifier` as they are not part of 8.3

mhsdesign commented on Neos.Neos/Classes/Command/WorkspaceCommandController.php:
```diff
@@ -178,13 +175,22 @@ public function createCommand(
     ): void {
         $contentRepositoryId = ContentRepositoryId::fromString($contentRepository);
+
+        $this->workspaceService->createPersonalWorkspace(
```
i think this is wrong, because we still use `$contentRepositoryInstance->handle(CreateWorkspace::create(` below

bwaidelich replied:
Right, the command controller needs some more love (and way less domain logic).
Good call, I'm on it!

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -266,17 +240,17 @@ public function editAction(WorkspaceName $workspaceName): void
-            $this->domainUserService->currentUserCanTransferOwnershipOfWorkspace($workspace)
-        );
+
+        // TODO fix $this->userService->currentUserCanTransferOwnershipOfWorkspace($workspace)
```
new todo here :D

bwaidelich replied:
The whole UI part is unfinished

mhsdesign commented on Neos.Workspace.Ui/Resources/Private/Templates/Workspace/Index.html:
```diff
@@ -14,146 +25,137 @@
+					<td>
+						<f:security.ifAccess privilegeTarget="Neos.Neos:Backend.Module.Administration.Users">
+							<f:then>
+								<neos:link.module path="administration/users" action="show" arguments="{user: workspace.workspaceOwner}">TODO</neos:link.module>
+							</f:then>
+							<f:else>
+								TODO
```
two todos hiding here as well :D

mhsdesign reviewed (changes requested):
Soo many `TODO` thanks, i just highlighted the 3 new ones introduced in chance any of them are by accident.

kitsunet reviewed (approved):
I do like this a lot, tried the migration on a couple of datasets, works fine. I am giving a preliminary approval despite failing tests. IMHO you two could merge this when you feel it's ready.

bwaidelich commented:
Thanks for your great help.
One comment: We currently throw if workspace metadata does not exist. We could instead always return metadata (if no record exists it should _not_ be created on the fly but we could just return a metadata dto with title == workspace name).
It would never have an owner but admins could still see/manage things.
Actually it was like this between some commits but I reverted because the mutations are more complex (upsert instead of update) with this.. But it _might_ be less error prone.. not sure

bwaidelich commented:
Thanks for your great help.
One comment: We currently throw if workspace metadata does not exist. We could instead always return metadata (if no record exists it should _not_ be created on the fly but we could just return a metadata dto with title == workspace name).
It would never have an owner but admins could still see/manage things.
Actually it was like this between some commits but I reverted because the mutations are more complex (upsert instead of update) with this.. But it _might_ be less error prone.. not sure

mhsdesign commented:
Jup that was also what i outlined [here](https://github.com/neos/neos-development-collection/pull/5146#pullrequestreview-2340553426) but the main problem as said would still persist: if "live" has no metadata no one can publish to it. So i guess we need a command to patch metadata for workspaces ... something like `workspace:chmod` :D

bwaidelich commented:
> if "live" has no metadata no one can publish to it

Just to avoid confusion: it wouldn't need metadata, but an entry in the `role` table.

An admin/CLI user could always add this and our migration should make sure that we have at least
```sql
INSERT INTO `neos_neos_workspace_role` (`content_repository_id`, `workspace_name`, `subject_type`, `subject`, `role`)
VALUES
	('default', 'live', 'GROUP', 'Neos.Neos:LivePublisher', 'COLLABORATOR');
```

bwaidelich committed:
Merge branch '9.0' into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+                    OR
+                    (subject_type = :groupSubjectType AND subject IN (:groupSubjects))
+                )
+            ORDER BY
+                CASE
+                    WHEN role='MANAGER' THEN 1
+                    WHEN role='COLLABORATOR' THEN 2
+                END
+            LIMIT 1
```
ordering doesnt make sense when we limit it to 1 right?

bwaidelich replied:
yes, it does because we only want to return the most specific role, i.e. if you're `MANAGER` _and_ `COLLABORATOR` we want to return `MANAGER` and not `COLLABORATOR` :)

mhsdesign replied:
marvellous i havent thought about this;)

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+                content_repository_id = :contentRepositoryId
+                AND workspace_name = :workspaceName
+                AND (
+                    (subject_type = :userSubjectType AND subject = :userId)
```
the whole part of a user id in the workspace_role table is outdated i think. At least in `createPersonalWorkspace` we never create an entry in this table, and we use the userId stored in the metadata for determining the permissions.

bwaidelich replied:
it's not outdated, I just removed the `OWNER` role and made that part of the metadata because that had some weird implications.
But it should still be possible to grant a single user specific permissions on a ws

mhsdesign replied:
ahh i see i was just confused that there is no write api for it ... just that we query this ;)
Thanks for thinking two steps ahead! ❤️

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+                AND (
+                    (subject_type = :userSubjectType AND subject = :userId)
+                    OR
+                    (subject_type = :groupSubjectType AND subject IN (:groupSubjects))
```
except for the migration we dont create any entries here, meaning new shared or root workspaces are not handed a `WorkspaceRole` and thus are not writeable and readable.

bwaidelich replied:
the `WorkspaceService` still misses some methods as discussed :)

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -0,0 +1,365 @@
+    /**
+     * Create a new, personal, workspace for the specified user if none exists yet
+     */
+    public function createPersonalWorkspaceForUserIfMissing(ContentRepositoryId $contentRepositoryId, User $user): void
```
the `void` signature here and the throwing at `getPersonalWorkspaceForUser` makes this not easily usable id say.

A future requirement of the Neos ui will be to get the content graph for the current user. For that we will have to figure out if a personal workspace already exists or create a new one based on live.

The code with the current api would be following:

```php
$service->createPersonalWorkspaceForUserIfMissing($cr, $user);
$workspace = $service->getPersonalWorkspaceForUser($cr, $user->getId());
$contentGraph = $cr->getContentGraph($workspace->workspaceName);
```

there are 3 things that i noticed:
- `$user->getId()` vs `$user`
- we need to fetch the workspace name via `findPrimaryWorkspaceNameForUser` twice and just discard the first result (could be resolved by returning it)
- we technically dont need to fetch the workspace at all but only its name to get the content graph (especially if fetching workspaces will be slightly more expensive)

bwaidelich replied:
> $user->getId() vs $user

what about it?

> we technically dont need to fetch the workspace at all but only its name to get the content graph (especially if fetching workspaces will be slightly more expensive)

isn't that a micro optimization for some method that is almost never called?

mhsdesign replied:
> what about it?

we should be consistent of either passing the whole user vs id inside one class i think ^^ but thats just a nitpick

> isn't that a micro optimization for some method that is almost never called?

dont worry i only expressed my concerns here because it is called on every call in the neos ui backend `neos/content` so ...

bwaidelich replied:
> we should be consistent of either passing the whole user vs id inside one class i think

Ah, gotcha.
I pass the whole user object because we need the user's _name_ too, to create a speaking title (see https://github.com/neos/neos-development-collection/blob/88e344d2068a0ef451cde56725a7960013fe2a51/Neos.Neos/Classes/Domain/Service/WorkspaceService.php#L144)

> dont worry i only expressed my concerns here because it is called on every call in the neos ui backend

but the expensive part would only be executed if no primary workspace was found, right?!

mhsdesign replied:
1. ahh i see sorry oversaw that ^^

2. yes the creation - the expensive part - only happens once, i was just thinking about how to optimise the content graph loading for when opening the backen but it probably doesnt matter and is a sidetrack. Premature optimisation etc... so lets go down the route as drawn out and see if we want to optimise it later ;)

mhsdesign replied:
1. ahh i see sorry oversaw that ^^

2. yes the creation - the expensive part - only happens once, i was just thinking about how to optimise the content graph loading for when opening the backen but it probably doesnt matter and is a sidetrack. Premature optimisation etc... so lets go down the route as drawn out and see if we want to optimise it later ;)

mhsdesign commented on Neos.Neos/Classes/Command/WorkspaceCommandController.php:
```diff
@@ -178,13 +175,22 @@ public function createCommand(
         } else {
             $workspaceOwnerUserId = UserId::fromString($owner);
-            $workspaceOwner = $this->userService->findByUserIdentifier($workspaceOwnerUserId);
+            $workspaceOwner = $this->userService->findUserById($workspaceOwnerUserId);
```
as discussed with basti, we want to fix this to use `getUser` like in Neos8.3 instead so the more human readable name is used again not the persistence object identifier.

mhsdesign reviewed (commented):
sorry :O

bwaidelich committed:
Rename `WorkspaceSubjectType` to `WorkspaceRoleSubjectType`

bwaidelich committed:
Make `WorkspaceService` more explicit

bwaidelich committed:
Workspace UI: Workspace role preset

Automatically set the current user as `Manager` and role `Neos.Neos:AbstractEditor` as `Collaborator` for new shared workspaces

bwaidelich committed:
`WorkspacePermissions` doc comments

bwaidelich committed:
Add more doc comments

bwaidelich committed:
wip

mhsdesign reviewed (approved):
Everything looks really good and perfectly thought out now thanks!
I also was able to remove the migration hack from the Neos ui e2e tests and use the assignRole command now.

The UI e2e tests mostly (except the flaky ones) pass and i just triggered one final run in the CI.

The Neos CI here is a bit kaput due to the Neos UI accidentally also being installed when testing and that crashes it due to a missing class and flow DI. See https://github.com/neos/neos-development-collection/issues/4951

Locally phpstan and the CR behat tests pass which is IMO a good enough indicator. Post merge well see if were right *g

bwaidelich committed:
Fix and improve Behat tests for WorkspaceService

...and fix the discovered bugs along the way o.O

bwaidelich committed:
Remove `workspaceName` from `WorkspaceMetadata`

bwaidelich committed:
Fix typo in private method

mhsdesign committed:
TASK: Improve documentation

bwaidelich committed:
Merge branch 'feature/4726-extract-workspace-metadata-and-user-assignment-to-neos' of https://github.com/neos/neos-development-collection into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

bwaidelich committed:
Allow to determine Workspace role assignments

bwaidelich committed:
Allow to determine Workspace role assignments

bwaidelich committed:
stabilize WorkspaceMetadata constructor

bwaidelich committed:
Fix `WorkspaceService::getWorkspacePermissionsForUser()`

bwaidelich committed:
Merge branch 'feature/4726-extract-workspace-metadata-and-user-assignment-to-neos' of https://github.com/neos/neos-development-collection into feature/4726-extract-workspace-metadata-and-user-assignment-to-neos

mhsdesign committed:
TASK: Use `WorkspaceRoleAssignment` in `assignWorkspaceRole` signatur
